src/utils/with_options/src/lib.rs

```diff
@@ -14,7 +14,34 @@
 
 use proc_macro::TokenStream;
 
-/// Annotates that the struct represents the WITH properties for a connector.
+/// Annotates that the struct represents the `WITH` properties for a connector.
+/// This is a dummy marker that has no effect on the generated code. It's used to generate the `with_option_*.yaml` files.
+///
+/// ## Notes about how to define property structs
+///
+/// ### Prefer strongly-typed fields
+///
+/// Avoid using `HashMap` with `#[serde(flatten)]` to include all unknown fields.
+/// Declare all fields explicitly instead.
+///
+/// The only exception now is CDC, which needs to pass a lot of options as-is to Debezium.
+///
+/// ### `serde` attributes
+///
+/// Always add `#[serde(deny_unknown_fields)]` to the top-level struct.
+/// However, do not add it if there's a `HashMap` field with `#[serde(flatten)]`.
+///
+/// ### Common struct
+///
+/// When there are some fields can be grouped together, and/or can be shared by source and sink,
+/// or by multiple connectors, define a struct for them and use the common struct as a
+/// `#[serde(flatten)]` field.
+///
+/// Add `#[derive(WithOptions)]` to both the outer and the inner struct.
```

Comment:
Can we make `WithOptions` a dummy trait to enforce that?

Reply:
What do you mean? Isn't it already a dummy trait?

Reply:
It's a dummy `derive` but not a dummy `trait`. 😄 I mean it can function like `std::fmt::Debug`, which will generate a function that call the same function on every field. So if the inner struct does not `impl std::fmt::Debug`, then compilation fails.

Reply:
https://github.com/risingwavelabs/risingwave/pull/14175